bpo-27820: Fix AUTH LOGIN logic in smtplib.SMTP

pepoluan opened the pull request:
Previously, `smtplib.SMTP.auth_login` responded with username if challenge is not given, and _always_ responded with password if challenge is given. This causes auth failure when `initial_response_ok=False`.

The standard for AUTH LOGIN specifies that clients should simply respond with base64-encoded username on first challenge, and with base64-encoded password on second challenge, disregarding the actual contents of the challenges themselves.

Suitable test cases have been added to `test_smtplib.py` + a fix for a long-standing bug in `test_smtplib.py` where setting `initial_response_ok=False` will always result in "Internal Confusion" error.


<!-- issue-number: [bpo-27820](https://bugs.python.org/issue27820) -->
https://bugs.python.org/issue27820
<!-- /issue-number -->

pepoluan committed:
Fix auth_login logic (bpo-27820)

pepoluan committed:
Add test for bpo-27820 fix

And in the process also fix a longstanding bug in the
SimSMTPChannel.found_terminator() method that causes inability to test
SMTP AUTH with initial_response_ok=False.

pepoluan commented on Lib/test/test_smtplib.py:
```diff
@@ -785,7 +785,7 @@ def found_terminator(self):
                 self.smtp_state = self.COMMAND
                 self.push('%s %s' % (e.smtp_code, e.smtp_error))
-                return
+            return
```
The mis-indented return caused the test for smtpd state to "fall through" to the next test, resulting ultimately in "Internal Confusion" error.

pepoluan commented on Lib/smtplib.py:
```diff
@@ -662,7 +666,7 @@ def auth_plain(self, challenge=None):
         """ Authobject to use with LOGIN authentication. Requires self.user and
         self.password to be set."""
-        if challenge is None:
+        if challenge is None or self._auth_challenge_count < 2:
```
The [draft standard for AUTH LOGIN](https://tools.ietf.org/html/draft-murchison-sasl-login-00) specifies that [clients SHOULD ignore](https://tools.ietf.org/html/draft-murchison-sasl-login-00#section-2.1) the actual contents of the challenge, and respond only based on how many challenges have been received.

orsenthil commented on Lib/smtplib.py:
```diff
@@ -633,10 +634,13 @@ def auth(self, mechanism, authobject, *, initial_response_ok=True):
+            self._auth_challenge_count = 0
         # If server responds with a challenge, send the response.
-        if code == 334:
+        while code == 334:
```
This `while` statement is unguarded. What if server keeps sending the challenge, wont this code block the client indefinitely?

gst replied:
it would be a broken/buggy server I believe..  against that you/we can protect with a loop count eventually. but what count to choose when to explode ?

pepoluan replied:
There's already `self._auth_challenge_count` there, so we can break on that.

That begets 2+1 questions: (1) What's the sensible limit, (1a) will we allow customization of the limit, (2) How to break? Just break or raise something?

pepoluan replied:
I've implemented the guard in 481ca55

I think this conversation is now resolved.

pepoluan reviewed (commented):
Response to some questions.

pepoluan committed:
Implement challenge count limit

Broken/buggy server might 'trap' us in AUTH.

We raise an exception and let the user handle things.

pepoluan commented on Lib/smtplib.py:
```diff
@@ -633,14 +635,23 @@ def auth(self, mechanism, authobject, *, initial_response_ok=True):
             response = encode_base64(
                 authobject(challenge).encode('ascii'), eol='')
             (code, resp) = self.docmd(response)
+            # If server keeps sending challenges, something is wrong.
+            if self._auth_challenge_count > _MAXCHALLENGE:
+                raise SMTPException(
+                    "Server AUTH mechanism infinite loop. Last response: "
+                    + repr((code, resp))
+                )
```
This guards against broken/buggy server that traps us in the AUTH state.

At this point, state is indeterminate, and we should let user handle things. That's why it simply raises. (The only "guaranteed safe" thing to do is `close()`, actually.)

pepoluan committed:
Add NEWS entry for bpo-27820 fix

orsenthil commented on Lib/smtplib.py:
```diff
@@ -64,6 +64,7 @@
 CRLF = "\r\n"
 bCRLF = b"\r\n"
 _MAXLINE = 8192 # more than 8 times larger than RFC 821, 4.5.3
+_MAXCHALLENGE = 5  # Maximum number of AUTH challenges sent
```
This looks good.  A test case to assert this this raise will make this complete. Would you like to add this, @pepoluan ?

Thank you!

pepoluan replied:
No problem. I'll make a test case and push the commit.

pepoluan replied:
Test case added in 08b9d63

pepoluan committed:
Add test case for AUTH trap

"AUTH trap" is when server keeps sending AUTH Challenge (334 ...)

pepoluan commented on Lib/test/test_smtplib.py:
```diff
@@ -851,6 +851,11 @@ def _auth_login(self, arg=None):
             self._authenticated(self._auth_login_user, password == sim_auth[1])
             del self._auth_login_user
 
+    def _auth_buggy(self, arg=None):
+        # This AUTH mechanism will 'trap' client in a neverending 334
+        # base64 encoded 'BuGgYbUgGy'
+        self.push('334 QnVHZ1liVWdHeQ==')
```
Totally made-up AUTH mechanism whose purpose is to 'trap' the client in an infinite loop of `334` (AUTH Challenge)

pepoluan commented on Lib/test/test_smtplib.py:
```diff
@@ -1069,6 +1074,44 @@ def testAUTH_LOGIN(self):
+            self.assertEqual(b"BuGgYbUgGy", challenge)
+            return "\0"
+
+        smtp = smtplib.SMTP(
+            HOST, self.port, local_hostname='localhost',
+            timeout=support.LOOPBACK_TIMEOUT
+        )
+        try:
+            smtp.user, smtp.password = sim_auth
+            smtp.ehlo("test_auth_buggy")
+            expect = r"^Server AUTH mechanism infinite loop.*"
+            with self.assertRaisesRegex(smtplib.SMTPException, expect) as cm:
+                smtp.auth("BUGGY", auth_buggy, initial_response_ok=False)
+        finally:
+            smtp.close()
```
Can't use smtplib.SMTP as context manager here ... its `__exit__` causes a different exception to be raised in addition. So the test uses a decidedly "lower level" mechanism to close the connection.

pepoluan committed:
Fix whitespace

VSCode did not indicate any problems, that's why I missed this
during previous push.

orsenthil reviewed (approved):
LGTM.   Thank you, @pepoluan .
